Obs wafer band in obsdb

msilvafe opened the pull request:
This is an update to obsdb to allow storing information keyed by obs-wafer-band without breaking the current obs-id primary key behavior and allowing us to retrieve common entries for an obs-id between 2 databases that could be keyed by obs-wafer-band or obsid.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -65,6 +80,11 @@ def __init__(self, map_file=None, init_db=True):
+        global TABLE_DEFS
+        global TABLE_DEFS_OWB
+        if obs_wafer_band:
+            TABLE_DEFS = TABLE_DEFS_OWB
```
This will modify the module level global var, affecting any future instantiations of ObsDb.  So instead you'll want to store the table_defs as an instance variable, and have all the internal functions refer to that. (`self._table_defs = TABLE_DEFS` / `TABLE_DEFS_OWB`).

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -160,22 +211,33 @@ def update_obs(self, obs_id, data={}, tags=[], commit=True):
         """
+        if (len(self.primary_keys) == 3):
+            wafer_band = self.warn_primary_keys(wafer_band)
+            obs_id = {'obs_id': obs_id, 'wafer_slot': wafer_band[0],
+                      'bandpass': wafer_band[1]}
+        else:
+            obs_id = {'obs_id': obs_id}
```
Instead of re-using `obs_id`, call this aggregate var something like `obs_key`.

That's a terminology we could adopt globally.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -314,11 +392,41 @@ def query(self, query_text='1', tags=None, sort=['obs_id'], add_prefix=''):
             results.keys = [add_prefix + k for k in results.keys]
         return results
+    
+    def query_linked_dbs(primary_db, secondary_db, query_condition,
+                         query_values=(), add_prefix=''):
```
This function is odd to see in the present context because it doesn't pay much attention to the 3-parameter keys... I guess it's intended for primary=normal_obsdb, and secondary=an_owb_obsdb.  Then you get the usual per-obs stuff, and an enhanced wafer/band entry for each obs.  Ok.  Might be worth a comment that either or neither of the dbs can be an owb.

Also `primary_db` is in the "self" position, so call it self.  Or make this a staticmethod.

Why only primary and secondary?  You could accept an arb number of dbs from the user.  (Though the primary db query is not the slow part -- the follow up, looping over obs for (each) secondary db will be.)  That being said, a helper function that accepts "primary_results", and looks up all this stuff in secondary, could be helpful; that's what you chain together to get primary -> secondary, thirdary, etc.

Instead of return list of dict of resultset.... easier to just return list of list/tuple of resultset?

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -81,10 +101,38 @@ def __init__(self, map_file=None, init_db=True):
+            warn_str = 'WARNING: Primary key(s)'
+            for i, wb in enumerate(wafer_band):
+                if wb is None:
+                    wafer_band[i] = '_all'
```
This failed for me because I passed in a tuple.  Also it's not great to modify someone's inputs if you don't need to.  Instead, start the func with `wafer_band = list(wafer_band)`; then you'll definitely be able to change it and return it.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -81,10 +101,38 @@ def __init__(self, map_file=None, init_db=True):
+            are not specified and ObsDb is indexed by obs_id, wafer_slot,
+            and bandpass. These keys will be set to _all.
+            """
+            print(warn_str)
```
use warnings.warn or whatever.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -81,10 +101,38 @@ def __init__(self, map_file=None, init_db=True):
+            raise ValueError("Unsupported number of primary keys in the 'obs' table"
+                             " only 1 or 3 are allowed.")
+
+    def get_primary_keys(self):
```
This conflicts with database terminology "primary key", which is the combination of the fields.  So a better name for this function is "get_primary_key_keys" or "get_primary_key_fields".  So how about "get_primary_fields".

Indeed the obsdb code is pretty rigorous in referring to the obsdb columns as "fields", rather than "keys".  So let's keep doing that.

mhasself reviewed (changes requested):
This is totally the right direction for obsdb, thanks for figuring it out.

Obviously this needs tests and stuff :) so there will be a few rounds of iteration.  But I'm on-board with the concept.

You've hard-coded the extension to OWB, which makes sense for SO.  But it's a small step to _fully_ generalize it for more arbitrary index combos.  So I might push for that in next iteration.  (The "get_primary_keys" function shows that you can figure out the obs_key for a given database.)

msilvafe commented:
> This is totally the right direction for obsdb, thanks for figuring it out.
> 
> Obviously this needs tests and stuff :) so there will be a few rounds of iteration. But I'm on-board with the concept.
> 
> You've hard-coded the extension to OWB, which makes sense for SO. But it's a small step to _fully_ generalize it for more arbitrary index combos. So I might push for that in next iteration. (The "get_primary_keys" function shows that you can figure out the obs_key for a given database.)

Ok @mhasself I think I've addressed all of your comments, please have another look when you can. Thanks.

msilvafe commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -93,42 +84,48 @@ def __init__(self, map_file=None, init_db=True, obs_wafer_band=False):
         return primary_keys
     
-    def warn_primary_keys(self, wafer_band):
+    def warn_primary_keys(self, wafer_info):
```
@mhasself what structure are you thinking that users could pass in a dict for this wafer_info?
Like:
```
{`wafer_slot':'ws0', 'wafer.bandpass':'f150'}
```
And I would return `['wafer_slot', 'bandpass']` or something like: `{wafer_info:['wafer_slot', 'bandpass']}`? I guess I'm not really sure what the useful aspect of the values in the key-value pairs of a dict would be in this case.

msilvafe commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -211,35 +208,38 @@ def update_obs(self, obs_id, data={}, tags=[],
-                      'bandpass': wafer_band[1]}
-        else:
-            obs_id = {'obs_id': obs_id}
+        obs_key = {'obs_id': obs_id}
+        if (len(self.primary_keys) > 1):
+            wafer_info = self.warn_primary_keys(wafer_info)
+            for i, k in enumerate(self.primary_keys[1:]):
+                obs_key[k] = wafer_info[i]
```
I see this is also a similar comment about generalizing to a dict structure as you added under the `warn_primary_keys` function. Can you please write out an example of what the `obs_id` as str + `wafer_info` as dict, or `obs_id` as dict + `wafer_info` as list of str  or `obs_id` as dict and `wafer_info` as dict, etc. would actually look like?

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -92,39 +92,63 @@ def __init__(self, map_file=None, init_db=True, wafer_info=None):
+    def _convert_wafer_info(self, obs_id, wafer_info):
+        """Helper function to allow flexibility in way obs_id and wafer_info are passed in."""
+        if isinstance(wafer_info, dict):
+            wafer_info = wafer_info.values()
```
I think you need to grab these by key -- they might not be in the order you expect.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -92,39 +92,63 @@ def __init__(self, map_file=None, init_db=True, wafer_info=None):
+                raise ValueError(f"obs_id tuple must be of length {len(self.primary_keys)}")
+        if isinstance(obs_id, dict):
+            if len(obs_id) == len(self.primary_keys):
+                wafer_info = tuple([wi for ki, wi in obs_id.items() if ki != 'obs_id'])
```
Same here -- you're relying on the order of .items().

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -92,39 +92,63 @@ def __init__(self, map_file=None, init_db=True, wafer_info=None):
+        self.primary_keys = self.get_primary_fields(wafer_info)
 
-    def get_primary_fields(self):
+    def get_primary_fields(self, wafer_info=None):
```
I guess this should be a private function.  Seems like you cache `self.primary_keys` on init, so that's what users should access.

mhasself commented on sotodlib/core/metadata/obsdb.py:
```diff
@@ -204,13 +228,29 @@ def update_obs(self, obs_id, data={}, tags=[],
-        Returns:
-            self.
+        Example of ways to pass updates to obsdb when there are multiple primary keys.
+        1) obs_id as str and wafer_info as tuple::
```
Add a blank line above 1) so it renders properly ([docs build](https://github.com/simonsobs/sotodlib/pull/1159)).

mhasself commented on tests/test_obsdb.py:
```diff
@@ -145,13 +145,32 @@ def test_owb_primary_linked_query(self):
+        # Method 3
+        db.update_obs({'obs_id':'myobs0', 'wafer_slot':'ws5', 'bandpass':'f090'},
+                      data={'data1': 997})
+        result_m3 = db.get({'obs_id':'myobs0', 'wafer_slot':'ws5', 'bandpass':'f090'})
```
For at least one of these dict tests, pass the keys in the non-canonical order.

mhasself reviewed (changes requested):
Minor comments -- sorry for slow on this.
